chapters/errors.Rmd

```diff
@@ -308,54 +345,82 @@ What we *should* do is check that there are two columns separated by a comma,
 that the first column contains strings,
 and that the second is numerical.
 
-## Kinds of Errors {#errors-kind}
+> **Kinds of Errors**
+>
+> The "`if` then `raise`" approach is sometimes referred to as "look before you leap",
+> while the `try/except` approach obeys the old adage that
+> "it's easier to ask for forgiveness than permission".
+> The first approach is more precise,
+> but has the shortcoming that programmers can't anticipate everything that can go wrong when running a program,
+> so there should always be an `except` somewhere
+> to deal with unexpected cases.
+>
+> The one rule we should *always* follow is to check for errors as early as possible
+> so that we don't waste the user's time.
+> Few things are as frustrating as being told at the end of an hour-long calculation
+> that the program doesn't have permission to write to an output directory.
+> It's a little extra work to check things like this up front,
+> but the larger your program or the longer it runs,
+> the more useful those checks will be.
+
+## Testing Error Handling {#errors-testing}
```

Comment:
Maybe the `#errors-testing` section should go after the `#errors-messages` section? I say that because the error messages section flows quite nicely from the `#errors-exceptions` section (i.e. we've just written our own error message `OSError(f'{fname} is not a CSV file.'` and now need to consider how to write good error messages).

So my suggested order for the chapter is:
1. `#errors-exceptions`
2. `#errors-messages`
3. `#errors-testing`
4. `#errors-logging`

Reply:
:+1: